### PR TITLE
Added groupByKey - method to group collection by key

### PR DESCRIPTION
**Pull request by shameerc**
This will enable us to group nested collection using a key. For example 

```
$categories = [
   ['id' => 1, 'name' => 'Category 1', 'parent_id' => 0],
   ['id' => 2, 'name' => 'Category 2', 'parent_id' => 1],
   ['id' => 3, 'name' => 'Category 3', 'parent_id' => 1],
];
$grouped = Collection::from($categories)->groupByKey('parent_id');
```

Ref #7

### CODE REVIEW

**Comment by DusanKasan** [comment]
The groupByKey should only accept $key, so you can get rid of the getCallable method. 

Then in the actual implementation you can call filter (with isCollection($collection) and has($collection, $key))groupBy with grouping function of get($collection, $key) + the isCollection($collection) checking.. Take a look how the pluck method is refactored for guidance. 

Also, add the documentation/example to README.md.

**Comment by shameerc** [comment]
Thanks for the feedback I'm still looking at the `pluck`, will refactor the code soon and then you can merge it :)

File: tests/spec/CollectionSpec.php [hunk 1]
```diff
@@ -228,17 +228,24 @@ function it_can_group_by()
...
-        $collection->get('small')->toArray()->shouldReturn([['letter' => 'a', 'type' => 'small']]);
+        $collection->get('small')->toArray()->shouldReturn([
+            ['letter' => 'a', 'type' => 'small']
+        ]);
         $collection->get('caps')->toArray()->shouldReturn([
             ['letter' => 'A', 'type' => 'caps'],
             ['letter' => 'B', 'type' => 'caps']
         ]);
+
+        $collection = $this->groupByKey('types');
```
Comment (DusanKasan) [inline tests/spec/CollectionSpec.php:247]: There is no need for this variable, use the return value straight away.

**Comment by DusanKasan** [comment]
Everything looks fine. Fix the Scrutinizer notice i commented on and i'll merge.

**Comment by DusanKasan** [comment]
Thank you for the contribution :)

**Pull request merged**